Fixes #22 - Exception on draft_destroy when has_one association is nil

npafundi opened the pull request:
`self.send(association.name)` will return `nil` if the association is empty (there are no dependents).  Currently, if the association is `has_one` and the result is `nil`, it is still wrapped in an array.  When we try to access this `nil` element in the following `.each` block, an exception is thrown.

The changes to model.rb resolve this issue.  Singular dependents are only wrapped in an array if they exist (and `nil` isn't wrapped).

The `.each` block is also only called if the dependents exist.  So, whether a `has_one` or `has_many` returns `nil`, the block will be bypassed and we avoid any exceptions.

I added an `OnlyChild` class and association to the parents to easily test this fix.  Without the fix, many of the current tests will fail.  With the fix in place, the tests will run properly again.

Hope this is helpful, @chrisdpeters!  As always, let me know if you want something changed.

chrisdpeters commented:
Can you explain this? "Without the fix, many of the current tests will fail." Sorry, I may be reading it wrong. I run the tests on 0.3.2, and they all pass.

Looks like we will need a v0.3.4. Thanks for getting involved with helping me maintain these issues. And especially thanks for including tests with your pull requests!

npafundi commented:
No worries -- I had a tough time putting this one in words.  Some of the current tests will fail if a `has_one` association is added, e.g. to the `Parent`.

In this commit, I slightly changed the schema.  I added a migration and an OnlyChild class, and then updated `Parent
* I added a migration and changed the `Parent` class to contain `has_one :only_child`.
If I put that in,

npafundi commented:
No worries -- I had a tough time putting this one in words.  Some of the current tests will fail if a `has_one` association is added, e.g. to the `Parent`.

In this commit, I slightly changed the schema.  I added a migration and an `OnlyChild` class, and then updated `Parent` to contain `has_one :only_child`.  With those changes (no changes to the specs directly), a number of them will fail.

The fix in this commit addresses schemas with `has_one` relationships, specifically those which don't currently have an association linked to them (so the association is nil).

If you clone this branch and reset the changes to model.rb, then migrate/test, you should see a lot of failing tests.  Hope that clears it up.  I can try to add more info too, if you'd like.

npafundi commented:
Hey @chrisdpeters, just wondering if you had a chance to look at this one yet.  No problem if you haven't, but if you want more explanation or info let me know.

chrisdpeters commented:
No, I haven't. I'll take a look sometime this week.

My mom passed away a couple weeks ago, so my priorities have been elsewhere until today. Glad to be back into the code though!

npafundi commented:
I'm very sorry, Chris.  No rush at all.

I may be submitting another PR this week as well, but I'm primarily working from my own fork.

chrisdpeters commented:
No problem, and thanks. I know you're excited about this, and I appreciate all of the bugs that you're fixing. Just wanted you to know why I needed to slow down for a little bit.
